lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,913 @@
+			}
+
+			return rpcResp, nil
+		},
+	)
+}
+
+// MuSig2Sign sends a MuSig2SignRequest to the remote signer and waits for the
+// corresponding response.
+//
+// NOTE: This is part of the RemoteSignerRequests interface.
+func (s *SignCoordinator) MuSig2Sign(_ context.Context,
+	in *signrpc.MuSig2SignRequest,
+	_ ...grpc.CallOption) (*signrpc.MuSig2SignResponse, error) {
+
+	req := &walletrpc.SignCoordinatorRequest_MuSig2SignRequest{
+		MuSig2SignRequest: in,
+	}
+
+	return processRequest(
+		s, 0,
+		func(reqId uint64) walletrpc.SignCoordinatorRequest {
+			return walletrpc.SignCoordinatorRequest{
+				RequestId:       reqId,
+				SignRequestType: req,
+			}
+		},
+		func(resp *signerResponse) (*signrpc.MuSig2SignResponse,
+			error) {
+
+			rpcResp := resp.GetMuSig2SignResponse()
+			if rpcResp == nil {
+				return nil, ErrUnexpectedResponse
+			}
+
+			return rpcResp, nil
+		},
+	)
+}
+
+// SignMessage sends a SignMessageReq to the remote signer and waits for the
+// corresponding response.
+//
+// NOTE: This is part of the RemoteSignerRequests interface.
+func (s *SignCoordinator) SignMessage(_ context.Context,
+	in *signrpc.SignMessageReq,
+	_ ...grpc.CallOption) (*signrpc.SignMessageResp, error) {
+
+	req := &walletrpc.SignCoordinatorRequest_SignMessageReq{
+		SignMessageReq: in,
+	}
+
+	return processRequest(
+		s, 0,
+		func(reqId uint64) walletrpc.SignCoordinatorRequest {
+			return walletrpc.SignCoordinatorRequest{
+				RequestId:       reqId,
+				SignRequestType: req,
+			}
+		},
+		func(resp *signerResponse) (*signrpc.SignMessageResp, error) {
+			signResp := resp.GetSignMessageResp()
+			if signResp == nil {
+				return nil, ErrUnexpectedResponse
+			}
+
+			return signResp, nil
+		},
+	)
+}
+
+// SignPsbt sends a SignPsbtRequest to the remote signer and waits for the
+// corresponding response.
+//
+// NOTE: This is part of the RemoteSignerRequests interface.
+func (s *SignCoordinator) SignPsbt(_ context.Context,
+	in *walletrpc.SignPsbtRequest,
+	_ ...grpc.CallOption) (*walletrpc.SignPsbtResponse, error) {
+
+	req := &walletrpc.SignCoordinatorRequest_SignPsbtRequest{
+		SignPsbtRequest: in,
+	}
+
+	return processRequest(
+		s, 0,
+		func(reqId uint64) walletrpc.SignCoordinatorRequest {
+			return walletrpc.SignCoordinatorRequest{
+				RequestId:       reqId,
+				SignRequestType: req,
+			}
+		},
+		func(resp *signerResponse) (*walletrpc.SignPsbtResponse,
+			error) {
+
+			signResp := resp.GetSignPsbtResponse()
+			if signResp == nil {
+				return nil, ErrUnexpectedResponse
+			}
+
+			return signResp, nil
+		},
+	)
+}
+
+// processRequest is a generic function that sends a request to the remote
+// signer and waits for the corresponding response. If a timeout is set, the
+// function will limit the execution time of the entire function to the
+// specified timeout. If it is not set, configured timeouts will be used for
+// the individual operations within the function.
+func processRequest[R any](
```

Comment:
Could make `R comparable`, then `extractResponse` could just be `extractResponse func(*signerResponse) R`.

Example:

```
		func(resp *signerResponse) *walletrpc.SignPsbtResponse {
			return resp.GetSignPsbtResponse()
		},


or

		func(resp *signerResponse) bool {
			return resp.GetPong()
		},

```

And `processRequest` could look like this:

```
func processRequest[R comparable](s *SignCoordinator, timeout time.Duration,
	generateRequest func(uint64) walletrpc.SignCoordinatorRequest,
	extractResponse func(*signerResponse) R) (R, error) {

	var zero R
	
	done, err := s.registerRequest()
	if err != nil {
		return zero, err
	}
	defer done()

	startTime := time.Now()

	// If a timeout is enforced, we will wait for the connection using the
	// specified timeout. Otherwise, we will wait for the connection using
	// the configured connection timeout.
	if timeout != 0 {
		err = s.waitUntilConnectedWithTimeout(timeout)
	} else {
		err = s.WaitUntilConnected()
	}

	if err != nil {
		return zero, err
	}

	reqID := s.nextRequestID.Add(1)
	req := generateRequest(reqID)

	cleanUpChannel := s.createResponseChannel(reqID)
	defer cleanUpChannel()

	log.Debugf("Sending a %t to the remote signer with request ID %d",
		req.SignRequestType, reqID)

	err = s.stream.Send(&req)
	if err != nil {
		return zero, err
	}

	var resp *walletrpc.SignCoordinatorResponse

	// If a timeout is enforced, we need to limit the entire execution time
	// of this function to the timeout. Therefore, we need to calculate the
	// remaining allowed execution time.
	// If no timeout is enforced, we will wait for the response using the
	// configured request timeout.
	if timeout != 0 {
		newTimeout := timeout - time.Since(startTime)

		if time.Since(startTime) > timeout {
			return zero, ErrRequestTimeout
		}

		resp, err = s.getResponse(reqID, newTimeout)
	} else {
		resp, err = s.getResponse(reqID, s.requestTimeout)
	}

	if err != nil {
		return zero, err
	}

	rpcResp := extractResponse(resp)
	if rpcResp == zero {
		return zero, ErrUnexpectedResponse
	}

	return rpcResp, nil
}
```

Reply:
Thanks a lot for further improvement suggestion :tada:! Implemented this in the latest push :)